x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -0,0 +1,158 @@
+    private final Logger logger;
+    private final RetrySettings retrySettings;
+    private final ThreadPool threadPool;
+    private final Executor executor;
+
+    public RetryingHttpSender(
+        Sender sender,
+        ThrottlerManager throttlerManager,
+        Logger logger,
+        RetrySettings retrySettings,
+        ThreadPool threadPool
+    ) {
+        this(sender, throttlerManager, logger, retrySettings, threadPool, threadPool.executor(UTILITY_THREAD_POOL_NAME));
+    }
+
+    // For testing only
+    RetryingHttpSender(
+        Sender sender,
+        ThrottlerManager throttlerManager,
+        Logger logger,
+        RetrySettings retrySettings,
+        ThreadPool threadPool,
+        Executor executor
+    ) {
+        this.sender = Objects.requireNonNull(sender);
+        this.throttlerManager = Objects.requireNonNull(throttlerManager);
+        this.logger = Objects.requireNonNull(logger);
+        this.retrySettings = Objects.requireNonNull(retrySettings);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.executor = Objects.requireNonNull(executor);
+    }
+
+    private class InternalRetrier extends RetryableAction<InferenceResults> {
+        private final HttpRequestBase request;
+        private final ResponseHandler responseHandler;
+
+        InternalRetrier(HttpRequestBase request, ResponseHandler responseHandler, ActionListener<InferenceResults> listener) {
+            super(
+                logger,
+                threadPool,
+                retrySettings.getSettings().initialDelay(),
+                retrySettings.getSettings().maxDelayBound(),
+                retrySettings.getSettings().timeoutValue(),
+                listener,
+                executor
+            );
+            this.request = request;
+            this.responseHandler = responseHandler;
+        }
+
+        @Override
+        public void tryAction(ActionListener<InferenceResults> listener) {
+            ActionListener<HttpResult> responseListener = ActionListener.wrap(result -> {
+                try {
+                    responseHandler.validateResponse(throttlerManager, logger, request, result);
+                    InferenceResults inferenceResults = responseHandler.parseResult(result);
+
+                    listener.onResponse(inferenceResults);
+                } catch (Exception e) {
+                    logException(request, result, responseHandler.getRequestType(), e);
+                    listener.onFailure(e);
+                }
+            }, e -> {
+                logException(request, responseHandler.getRequestType(), e);
+                listener.onFailure(transformIfRetryable(e));
+            });
+
+            sender.send(request, responseListener);
+        }
+
+        @Override
+        public boolean shouldRetry(Exception e) {
+            if (e instanceof RetryException) {
+                return ((RetryException) e).shouldRetry();
```

Comment:
If we leave this as a runtime exception we can just do the `instanceof` check I believe.

Reply:
Added to the TODO list to remove `shouldRetry`.